File: backend/app/seed.py
import json
import uuid
from pathlib import Path

import bcrypt

from .database import db, utc_now

ASSETS_DIR = Path(__file__).parent.parent.parent / "assets"


def seed_if_empty() -> None:
    with db() as conn:
        count = conn.execute("SELECT COUNT(*) FROM vendors").fetchone()[0]
        if count > 0:
            return

        vendors = json.loads((ASSETS_DIR / "vendors.json").read_text(encoding="utf-8"))
        products = json.loads((ASSETS_DIR / "products.json").read_text(encoding="utf-8"))
        pin_hash = bcrypt.hashpw(b"1234", bcrypt.gensalt()).decode("utf-8")
        created_at = utc_now()

        for vendor in vendors:
            conn.execute(
                """
                INSERT INTO vendors (
                  id, name, location, description, phone, whatsapp,
                  rating, review_count, lat, lng, pin_hash, created_at, updated_at
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                """,
                (
                    vendor["id"],
                    vendor["name"],
                    vendor["location"],
                    vendor.get("description", ""),
                    vendor["phone"],
                    vendor.get("whatsapp", vendor["phone"]),
                    vendor.get("rating", 0),
                    vendor.get("reviewCount", 0),
                    vendor.get("lat"),
                    vendor.get("lng"),
                    pin_hash,
                    created_at,
                    created_at,
                ),
            )

        for product in products:
            conn.execute(
                """
                INSERT INTO products (
                  id, vendor_id, name, price_cents, description, category, tags,
                  image_asset, in_stock, can_buy_online, created_at, updated_at
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                """,
                (
                    product["id"],
                    product["vendorId"],
                    product["name"],
                    product["priceCents"],
                    product.get("description", ""),
                    product.get("category", "Herbal remedy"),
                    json.dumps(product.get("tags", [])),
                    product.get("imageAsset"),
                    0 if product.get("inStock") is False else 1,
                    1 if product.get("canBuyOnline") else 0,
                    created_at,
                    created_at,
                ),
            )

        print(f"Seeded {len(vendors)} vendors and {len(products)} products.")
        print("Demo vendor PIN for seeded vendors: 1234")


def new_vendor_id() -> str:
    return f"vendor-{uuid.uuid4().hex[:8]}"


def new_product_id() -> str:
    return f"prod-{uuid.uuid4().hex[:8]}"


def new_order_id() -> str:
    return f"ORD-{uuid.uuid4().hex[:8].upper()}"


def new_review_id() -> str:
    return f"rev-{uuid.uuid4().hex[:8]}"

File: backend/src/seed.js
const fs = require('fs');
const path = require('path');
const bcrypt = require('bcryptjs');
const db = require('./db');

const assetsDir = path.join(__dirname, '..', '..', 'assets');
const now = () => new Date().toISOString();

function readJson(filename) {
  const filePath = path.join(assetsDir, filename);
  return JSON.parse(fs.readFileSync(filePath, 'utf8'));
}

function seed() {
  const vendorCount = db.prepare('SELECT COUNT(*) AS c FROM vendors').get().c;
  if (vendorCount > 0) {
    console.log('Database already seeded — skipping.');
    return;
  }

  const vendors = readJson('vendors.json');
  const products = readJson('products.json');
  const pinHash = bcrypt.hashSync('1234', 10);

  const insertVendor = db.prepare(`
    INSERT INTO vendors (
      id, name, location, description, phone, whatsapp,
      rating, review_count, lat, lng, pin_hash, created_at, updated_at
    ) VALUES (
      @id, @name, @location, @description, @phone, @whatsapp,
      @rating, @review_count, @lat, @lng, @pin_hash, @created_at, @updated_at
    )
  `);

  const insertProduct = db.prepare(`
    INSERT INTO products (
      id, vendor_id, name, price_cents, description, category, tags,
      image_asset, in_stock, can_buy_online, created_at, updated_at
    ) VALUES (
      @id, @vendor_id, @name, @price_cents, @description, @category, @tags,
      @image_asset, @in_stock, @can_buy_online, @created_at, @updated_at
    )
  `);

  const seedAt = now();
  const tx = db.transaction(() => {
    for (const vendor of vendors) {
      insertVendor.run({
        id: vendor.id,
        name: vendor.name,
        location: vendor.location,
        description: vendor.description || '',
        phone: vendor.phone,
        whatsapp: vendor.whatsapp || vendor.phone,
        rating: vendor.rating ?? 0,
        review_count: vendor.reviewCount ?? 0,
        lat: vendor.lat ?? null,
        lng: vendor.lng ?? null,
        pin_hash: pinHash,
        created_at: seedAt,
        updated_at: seedAt,
      });
    }

    for (const product of products) {
      insertProduct.run({
        id: product.id,
        vendor_id: product.vendorId,
        name: product.name,
        price_cents: product.priceCents,
        description: product.description || '',
        category: product.category || 'Herbal remedy',
        tags: JSON.stringify(product.tags || []),
        image_asset: product.imageAsset ?? null,
        in_stock: product.inStock === false ? 0 : 1,
        can_buy_online: product.canBuyOnline ? 1 : 0,
        created_at: seedAt,
        updated_at: seedAt,
      });
    }

    db.prepare(
      'INSERT OR REPLACE INTO meta (key, value) VALUES (?, ?)',
    ).run('catalog_seeded_at', seedAt);
  });

  tx();
  console.log(`Seeded ${vendors.length} vendors and ${products.length} products.`);
  console.log('Demo vendor PIN for all seeded vendors: 1234');
}

module.exports = { seed };

if (require.main === module) {
  seed();
}
